Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+#include "Python.h"
+
+#include "pycore_llist.h"
+#include "pycore_lock.h"        // _PyRawMutex
+#include "pycore_parking_lot.h"
+#include "pycore_pyerrors.h"    // _Py_FatalErrorFormat
+#include "pycore_pystate.h"     // _PyThreadState_GET
+
+#ifdef MS_WINDOWS
+#   define WIN32_LEAN_AND_MEAN
+#   include <windows.h>
+#elif (defined(_POSIX_SEMAPHORES) && !defined(HAVE_BROKEN_POSIX_SEMAPHORES) && \
+     defined(HAVE_SEM_TIMEDWAIT))
+#   define USE_SEMAPHORES
+#   include <semaphore.h>
+#elif defined(HAVE_PTHREAD_H)
+#   include <pthread.h>
+#elif defined(HAVE_PTHREAD_STUBS)
+#   include "cpython/pthread_stubs.h"
+#else
+#   error "Require native threads. See https://bugs.python.org/issue31370"
+#endif
+
+// A simple, cross-platform binary semaphore that can be used to implement
+// wakeup/sleep.
+struct _PySemaphore {
+#if defined(MS_WINDOWS)
+    HANDLE platform_sem;
+#elif defined(USE_SEMAPHORES)
+    sem_t platform_sem;
+#else
+    PyMUTEX_T mutex;
+    PyCOND_T cond;
+    int counter;
+#endif
+};
+
+typedef struct {
+    // The mutex protects the waiter queue and the num_waiters counter.
+    _PyRawMutex mutex;
+
+    // Linked list of `struct wait_entry` waiters in this bucket.
+    struct llist_node root;
+    size_t num_waiters;
+} Bucket;
+
+struct wait_entry {
+    struct _PyUnpark unpark;
+    uintptr_t addr;
+    _PySemaphore *sema;
+    struct llist_node node;
+};
+
+#define MAX_SEMA_DEPTH 3
+
+typedef struct {
+    Py_ssize_t refcount;
+
+    int depth;
+    _PySemaphore semas[MAX_SEMA_DEPTH];
+} ThreadData;
+
+#define NUM_BUCKETS 251
+
+// Table of waiters (hashed by address)
+static Bucket buckets[NUM_BUCKETS];
+
+#ifdef HAVE_THREAD_LOCAL
+static _Py_thread_local ThreadData *thread_data = NULL;
```

Comment:
It would be nice if we didn't have to use a pointer here:

```suggestion
static _Py_thread_local ThreadData thread_data = {0};
```

but we have to in order to init/destroy `thread_data->semas` properly, right?

Reply:
Yes, we would still need to init/destroy `thread_data->semas`. I'm also concerned about compatibility: our other `_Py_thread_local` variables are all pointers and I remember macOS having issues with thread-local aggregate types.

Reply:
I've removed the thread-local data.